bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/CharResponseWrapper.java

```diff
@@ -37,14 +39,21 @@ public class CharResponseWrapper extends HttpServletResponseWrapper {
     public CharResponseWrapper(HttpServletResponse response) {
         super(response);
         writer = new CharArrayWriter();
+        wrapped = response;
+
     }
 
     /**
      * Returns the newly created {@link PrintWriter} object created by this wrapper
      * @return - newly created {@link PrintWriter}
      */
     @Override
-    public PrintWriter getWriter() {
+    public PrintWriter getWriter() throws IOException {
+        // call getWriter() on the original response, as this is used to initialise the Sling Rewriter Pipeline
+        // if not called this way, the TableOfContentsFilter would call getWriter() on the upstream response first in the context of a
+        // cq:Page resource, and hence the resourceType condition to select the processor configuration does not work
+        wrapped.getWriter();
```

Comment:
Just a note here, something we also discussed direclty: getters should not produce this kind of side-effects. But it is what it is.

Reply:
yes unfortunately.